compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2043,42 +2081,61 @@ private boolean exprIncompatible(BType eType, BLangExpression expr, AnalyzerData
         return checkExpr(exprToCheck, eType, data) == symTable.semanticError;
     }
 
-    private List<BType> checkExprList(List<BLangExpression> exprs, SymbolEnv env, AnalyzerData data) {
-        return checkExprList(exprs, env, symTable.noType, data);
+    private InferredTupleDetails checkExprList(List<BLangExpression> exprs, AnalyzerData data) {
+        return checkExprList(exprs, symTable.noType, data);
     }
 
-    private List<BType> checkExprList(List<BLangExpression> exprs, SymbolEnv env, BType expType, AnalyzerData data) {
-        List<BType> types = new ArrayList<>();
+    private InferredTupleDetails checkExprList(List<BLangExpression> exprs, BType expType, AnalyzerData data) {
+        InferredTupleDetails inferredTupleDetails = new InferredTupleDetails();
         SymbolEnv prevEnv = data.env;
         BType preExpType = data.expType;
-        data.env = env;
         data.expType = expType;
         for (BLangExpression e : exprs) {
             if (e.getKind() == NodeKind.LIST_CONSTRUCTOR_SPREAD_OP) {
                 BLangExpression spreadOpExpr = ((BLangListConstructorSpreadOpExpr) e).expr;
                 BType spreadOpExprType = checkExpr(spreadOpExpr, expType, data);
-                types.addAll(getListConstSpreadOpMemberTypes(e.pos, spreadOpExprType));
+                updateInferredTupleDetailsFromSpreadMember(e.pos, spreadOpExprType, inferredTupleDetails);
                 continue;
             }
 
             checkExpr(e, expType, data);
-            types.add(data.resultType);
+            if (inferredTupleDetails.restMemberTypes.isEmpty()) {
+                inferredTupleDetails.fixedMemberTypes.add(data.resultType);
+            } else {
+                inferredTupleDetails.restMemberTypes.add(data.resultType);
+            }
         }
         data.env = prevEnv;
         data.expType = preExpType;
-        return types;
+        return inferredTupleDetails;
+    }
+
+    private static class InferredTupleDetails {
```

Comment:
Can't we use only `InferredTuple` without the word `Details`?

Reply:
I thought keeping the name as just `InferredTuple` could be misleading.

Say we have,
`InferredTuple inferredTuple = new InferredTuple();`

So referring `inferredTuple` somewhere down in the code, one could think for a moment that we are referring to a `BLangTupleTypeNode` representing inferred tuple(as the name implies). Since it is not a `BLangTupleTypeNode` and we only keep type details that we need to construct the inferred tuple, felt something like `InferredTupleDetails` is more appropriate. WDYT?

Reply:
That makes sense.